Plugins/nosReflect/Source/Variables.cpp

```diff
@@ -0,0 +1,277 @@
+#include <Nodos/PluginHelpers.hpp>
+
+#include <shared_mutex>
+
+#include "Names.h"
+
+namespace nos::reflect
+{
+NOS_REGISTER_NAME(SetVariable)
+NOS_REGISTER_NAME(GetVariable)
+
+struct VariableContainer
+{
+	void Associate(std::string const& name, nosUUID const& node)
+	{
+		std::unique_lock lock(VariablesMutex);
+		auto it = Users.find(name);
+		if (it != Users.end())
+			it->second.insert(node);
+		else
+			Users.insert({name, {node}});
+	}
+
+	void Disassociate(std::string const& name, nosUUID const& node)
+	{
+		std::unique_lock lock(VariablesMutex);
+		auto it = Users.find(name);
+		if (it != Users.end())
+		{
+			auto& users = it->second;
+			users.erase(node);
+			if (users.empty())
+			{
+				nosEngine.LogI("Erasing variable %s", it->first.c_str());
+				Variables.erase(it->first);
+				Types.erase(it->first);
+				Users.erase(it);
+				UpdateStrings();
+			}
+		}
+	}
+
+	void SetVariable(std::string const& name, nos::Buffer const& value)
+	{
+		std::unique_lock lock(VariablesMutex);
+		bool update = !Variables.contains(name);
+		Variables[name] = value;
+		if (update)
+			UpdateStrings();
+		for (auto const& user : Users[name])
+		{
+			nosEngine.SetPinValue(user, value);
+		}
+	}
+
+	bool GetVariable(std::string const& name, nos::Buffer& outCopy)
+	{
+		std::shared_lock lock(VariablesMutex);
+		auto it = Variables.find(name);
+		if (it != Variables.end())
+		{
+			outCopy = nos::Buffer(it->second);
+			return true;
+		}
+		return false;
+	}
+
+	void NotifyUsersAboutTypeChange(std::string const& name, nos::Name const& typeName)
+	{
+		for (auto const& user : Users[name])
+		{
+			NodeContext::SetPinType(user, typeName);
+		}
+	}
+
+	void UpdateStrings()
+	{
+		auto names = GetAllVariableNames();
+		UpdateStringList("nos.reflect.VariableNames", names);
+		std::stringstream s;
+		s << "List of variables:\n";
+		for (auto& name : names)
+			s << name << '\n';
+		nosEngine.LogDI(s.str().c_str(), "Variable list updated");
+	}
+
+	std::vector<std::string> GetAllVariableNames()
+	{
+		std::vector<std::string> names;
+		names.reserve(Variables.size());
+		for (auto const& [name, value] : Variables)
+		{
+			names.push_back(name);
+		}
+		return names;
+	}
+
+	void SetType(std::string const& name, nos::Name const& typeName)
+	{
+		std::unique_lock lock(VariablesMutex);
+		Types[name] = typeName;
+		NotifyUsersAboutTypeChange(name, typeName);
+	}
+
+	std::optional<nos::Name> GetType(std::string const& name)
+	{
+		std::shared_lock lock(VariablesMutex);
+		auto it = Types.find(name);
+		if (it != Types.end())
+			return it->second;
+		return std::nullopt;
+	}
+	
+	std::unordered_map<std::string, nos::Buffer> Variables;
+	std::unordered_map<std::string, nos::Name> Types; // variable name -> type name
+	std::unordered_map<std::string, std::unordered_set<nosUUID>> Users;
```

Comment:
Merge all these maps into single unordered_map by creating a struct with nos;:Buffer, nos::Name and std::unordered_set<nosUUID> ?